text/0000-delegation.md

```diff
@@ -0,0 +1,700 @@
+
+impl AddOnlyCounter {
+    fn new() -> Self {
+        Self { 0 }
+    }
+}
+
+impl AddAssign for AddOnlyCounter {
+    delegate * to 0;
+}
+```
+
+By using the _newtype pattern_, we _restrict_ the behaviour of the original type and only allow semantically valid methods/operations. If `AddOnlyCounter` is defined in another module, users can't break its guarantees, as the `u32` is private aka encapsulated.
+
+```rust
+let counter = AddOnlyCounter::new();
+
+// this works:
+counter += 1;
+
+// these are compile-time errors:
+counter = 5;
+counter -= 1;
+counter.0 -= 1;
+```
+
+
+# Reference-level explanation
+[reference]: #reference
+
+A delegation statement can only appear inside a trait impl block. Delegation inside inherent impls is left as a future extension.
+
+Delegation must be to a field on `Self`. Other kinds of implementer expressions are left as future extensions. This also means delegation can only be done on structs for now.
+
+There may be more than one delegation statement. For readability, `rustfmt` moves delegation statements to the top of an impl block.
+
+A delegation statement always consists of:
+
+-   the keyword `delegate`
+-   either a `*`, or a comma-separated list of items being delegated
+-   the contextual keyword `to`
+-   the delegation target `field_name`
+-   a semicolon
+
+An “item being delegated” is always two tokens. The first token must be either `fn`, `type` or `const`. The second is any valid identifier for a trait item.
+
+The semantics of a delegation statement should be the same as if the programmer had written each delegated item implementation manually. For instance, if the trait `TR` has a default implementation for method `foo()`, and the type `F` does not provide its own implementation, then delegating `TR` to `F` means using `TR`’s implementation of `foo()`. If `F` does provide its own implementation, then delegating `TR` to `F` means using `F`’s implementation of `foo()`. The only additional power granted by this feature is that `delegate *` can automatically change what items get implemented if the underlying trait `TR` and type `F` get changed accordingly. There can be at most one `delegate *` per `impl` block.
+
+To generate the wrapper function:
+
+-   The function signature is copied from the function being delegated to.
+-   The self parameter is mapped to the implementer expression `self.field_name` for any type `T` where `T: Deref<Target = Self>`, as per the [Custom Self Types RFC](https://github.com/rust-lang/rfcs/pull/2362). e.g.
+    ```rust
+    impl TR for S {
+        fn foo(self: Box<Self>) -> u32 {
+            self.f.foo()
+        }
+    }
+    ```
+-   `.trait_method_name()` is appended to the implementer expression.
+-   Subsequent parameters are passed through according to their type. e.g.
+    ```rust
+    fn check_name(&self, name: &str, ignore_capitals: bool, state: &mut State) -> bool {
+        self.f.check_name(&name, ignore_capitals, &mut state)
+    }    
```

Comment:
I don't think I know what "according to their type" means here.  Why are the extra `&`s needed, instead of just moving them all?  Perhaps `self.f.check_name({name}, {ignore_capitals}, {state})`.